magenta-lib/src/main/scala/magenta/input/DeploymentResolver.scala

```diff
@@ -1,68 +1,67 @@
-    if (missingDeps.nonEmpty)
-      Left(missingDeps.map(dep => s"Missing dependency $dep in $name"))
-    else {
-      deployment match {
-        case deployment @ DeploymentOrTemplate(Some(deploymentType), None, stacks, regions, _, _, _, _) =>
-          // terminating case - apply any default stacks and regions
-          val deploymentWithGlobalDefaults = deployment.copy(stacks = stacks.orElse(yaml.stacks), regions = regions.orElse(yaml.regions))
-          Right(deploymentWithGlobalDefaults)
+  def resolve(yaml: RiffRaffDeployConfig): List[Either[(String, String), Deployment]] = {
+    yaml.deployments.toList.map { case (label, rawDeployment) =>
+      for {
+        templated <- applyTemplates(label, rawDeployment, yaml.templates).right
+        deployment <- resolveDeployment(label, templated, yaml.stacks, yaml.regions).right
+      } yield deployment
+    }
+  }
 
-        case template @ DeploymentOrTemplate(None, Some(templateName), stacks, regions, app, contentDirectory, dependencies, parameters) =>
-          // template case - recursively resolve
-          for {
-            template <- yaml.templates.flatMap(_.get(templateName))
-              .toRight(List(s"Template with name $templateName (specified in $name) does not exist")).right
-            resolved <- resolveDeployment(templateName, template, yaml).right
-          } yield {
-            DeploymentOrTemplate(
-              resolved.`type`,
-              None,
-              stacks.orElse(resolved.stacks),
-              regions.orElse(resolved.regions),
-              app.orElse(resolved.app),
-              contentDirectory.orElse(resolved.contentDirectory),
-              dependencies.orElse(resolved.dependencies),
-              Some(resolved.parameters.getOrElse(Map.empty) ++ parameters.getOrElse(Map.empty))
-            )
-          }
-      }
+  /**
+    * Validates and resolves a templated deployment by merging its
+    * deployment attributes with any globally defined properties.
+    */
+  private[input] def resolveDeployment(label: String, templated: DeploymentOrTemplate, globalStacks: Option[List[String]], globalRegions: Option[List[String]]): Either[(String, String), Deployment] = {
+    for {
+      deploymentType <- templated.`type`.toRight(label -> "No type field provided").right
+      stacks <- templated.stacks.orElse(globalStacks).toRight(label -> "No stacks provided").right
+    } yield {
+      Deployment(
+        name              = label,
+        `type`            = deploymentType,
+        stacks            = stacks,
+        regions           = templated.regions.orElse(globalRegions).getOrElse(DEFAULT_REGIONS),
+        app               = templated.app.getOrElse(label),
+        contentDirectory  = templated.contentDirectory.getOrElse(label),
+        dependencies      = templated.dependencies.getOrElse(Nil),
+        // TODO? do we need to validate required parameters here,
```

Comment:
Totally mis-read that. No - I don't think we should validate deployment types or parameters here. At this stage we are just creating deployment types. We should validate that dependencies exist.